packages/squiggle-lang/__tests__/Reducer/Reducer_test.res

```diff
@@ -45,37 +59,47 @@ describe("eval", () => {
   // See https://mathjs.org/docs/expressions/syntax.html
   // See https://mathjs.org/docs/reference/functions.html
   describe("expressions", () => {
-    test("1", () => expectEvalToBe("1", "Ok(1)"))
-    test("1+2", () => expectEvalToBe("1+2", "Ok(3)"))
-    test("(1+2)*3", () => expectEvalToBe("(1+2)*3", "Ok(9)"))
-    test("2>1", () => expectEvalToBe("2>1", "Ok(true)"))
-    test("concat('a ', 'b')", () => expectEvalToBe("concat('a ', 'b')", "Ok('a b')"))
-    test("log(10)", () => expectEvalToBe("log(10)", "Ok(2.302585092994046)"))
-    test("cos(10)", () => expectEvalToBe("cos(10)", "Ok(-0.8390715290764524)"))
+    testEvalToBe("1", "Ok(1)")
+    testEvalToBe("1+2", "Ok(3)")
+    testEvalToBe("(1+2)*3", "Ok(9)")
+    testEvalToBe("2>1", "Ok(true)")
+    testEvalToBe("concat('a ', 'b')", "Ok('a b')")
+    testEvalToBe("log(10)", "Ok(2.302585092994046)")
+    testEvalToBe("cos(10)", "Ok(-0.8390715290764524)")
     // TODO more built ins
   })
   describe("arrays", () => {
     test("empty array", () => expectEvalToBe("[]", "Ok([])"))
-    test("[1, 2, 3]", () => expectEvalToBe("[1, 2, 3]", "Ok([1, 2, 3])"))
-    test("['hello', 'world']", () => expectEvalToBe("['hello', 'world']", "Ok(['hello', 'world'])"))
-    test("index", () => expectEvalToBe("([0,1,2])[1]", "Ok(1)"))
-    test("index not found", () =>
-      expectEvalToBe("([0,1,2])[10]", "Error(Array index not found: 10)")
-    )
+    testEvalToBe("[1, 2, 3]", "Ok([1, 2, 3])")
+    testEvalToBe("['hello', 'world']", "Ok(['hello', 'world'])")
+    testEvalToBe("([0,1,2])[1]", "Ok(1)")
+    testDescEvalToBe("index not found", "([0,1,2])[10]", "Error(Array index not found: 10)")
   })
   describe("records", () => {
     test("define", () => expectEvalToBe("{a: 1, b: 2}", "Ok({a: 1, b: 2})"))
     test("index", () => expectEvalToBe("{a: 1}.a", "Ok(1)"))
     test("index not found", () => expectEvalToBe("{a: 1}.b", "Error(Record property not found: b)"))
   })
+
+  describe("multi-line", () => {
+    testEvalToBe("1; 2", "Error(Assignment expected)")
+    testEvalToBe("1+1; 2+1", "Error(Assignment expected)")
+  })
+  describe("assignment", () => {
+    testEvalToBe("x=1; x", "Ok(1)")
+    testEvalToBe("x=1+1; x+1", "Ok(3)")
+    testEvalToBe("x=1; y=x+1; y+1", "Ok(3)")
+    testEvalToBe("1; x=1", "Error(Assignment expected)")
+    testEvalToBe("1; 1", "Error(Assignment expected)")
```

Comment:
This seems fine to me, I don't see why we'd want it to fail.

Reply:
assignments are returning bindings as a result  for the later expression's use. To allow expressions in the middle of code I have to manipulate expression to internally convert
"1; 1"
to
"_ = 1; 1"
, for example. (can be done as a feature)

If you say that you want use side effect creating statements in the middle). But why? You need to show me some examples sometime.
The interpretation is pure function application. There is no built in bindings management in the code.
And this code is nonsense
{
x=1;
y=2
}
Because y is discarded it is valid only in the block scope. Second as an assingment y=2 is returning bindings as block result not the 2. Then it would be Error(you are assigning unused variable y).

Is there a confusion? Are we making a functional language or just a sometimes functional like imparative script?

Reply:
Additional note. The philosophy of reducer is connecting a parser with the reducer using "functional semantics". Not implementing it per grammar. The semantic definition assignments is really a function that return bindings. But eventually any grammar is possible. Of course as long as it is a functional language. Not functionalish.

Semantic definitions are how you define the meaning of the parsed grammar.

`{1; 2;} `can become `{_ = 1; 2;}` via an additional semantic definition and it can work.
You can even make {1; y=2;} to work. In this case the semantic conversion would be
`{ _ = 1; y = 2; $$currentBindings[y] }`

Reply:
There are 2 possibilities
Defining a semantic rule that returns a record of all variables. So the code is auto converted to
```
proportion_of_funding_available = 0.7 to 0.8
total_funding_available = donation_size * proportion_of_funding_available
%{ proportion_of_funding_available, total_funding_available}
```
Or actually I would prefer your suggestion.